posthog/models/action.py

```diff
@@ -24,61 +24,54 @@ def calculate_events(self, start=None, end=None):
 
         try:
             if recalculate_all:
-                event_query, params = Event.objects.query_db_by_action(self).only("pk").query.sql_with_params()
+                event_queryset = Event.objects.query_db_by_action(self).only("pk")
             else:
-                event_query, params = (
-                    Event.objects.query_db_by_action(self, start=start, end=end).only("pk").query.sql_with_params()
-                )
-
+                event_queryset = Event.objects.query_db_by_action(self, start=start, end=end).only("pk")
+            event_query, params = event_queryset.query.sql_with_params()
         except EmptyResultSet:
-            self.last_calculated_at = calculated_at
-            self.is_calculating = False
-            self.save()
             self.events.all().delete()
-            return
-
-        query = """DELETE FROM "posthog_action_events" WHERE "action_id" = {}""".format(self.pk)
-
-        period_delete_query = """AND "event_id" in
-                       (SELECT id
-                        FROM posthog_event
-                        WHERE "created_at" >= '{}'
-                        AND "created_at" < '{}');
-                """.format(
-            start.isoformat(), end.isoformat()
-        )
-
-        insert_query = """INSERT INTO "posthog_action_events" ("action_id", "event_id")
-                        {}
-                    ON CONFLICT DO NOTHING
-                 """.format(
-            event_query.replace("SELECT ", "SELECT {}, ".format(self.pk), 1)
-        )
-
-        if not recalculate_all:
-            query += period_delete_query
         else:
-            query += ";"
-
-        query += insert_query
-
-        cursor = connection.cursor()
-        with transaction.atomic():
-            try:
-                cursor.execute(query, params)
-            except Exception as err:
-                capture_exception(err)
-
-        self.is_calculating = False
-        self.last_calculated_at = calculated_at
-        self.save()
+            delete_query = f"""DELETE FROM "posthog_action_events" WHERE "action_id" = {self.pk}"""
+
+            if not recalculate_all:
+                delete_query += f""" AND "event_id" IN (
+                    SELECT id FROM posthog_event
+                    WHERE "created_at" >= '{start.isoformat()}' AND "created_at" < '{end.isoformat()}'
+                )"""
+
+            insert_query = """INSERT INTO "posthog_action_events" ("action_id", "event_id")
+                            {}
+                        ON CONFLICT DO NOTHING
+                    """.format(
+                event_query.replace("SELECT ", f"SELECT {self.pk}, ", 1)
+            )
+
+            cursor = connection.cursor()
+            with transaction.atomic():
+                try:
+                    cursor.execute(delete_query + ";" + insert_query, params)
+                except Exception as err:
+                    capture_exception(err)
+
+            if self.post_to_slack:
```

Comment:
I think we only do this for periodic recalculations rather than full recalculations.

Scenario: User changes definition for "user signed up". With this change I think the user would suddenly get 20k notifications in slack.

Reply:
Hm, they should still only get notifications for events since last calculation, since the filter below has that condition

Reply:
If a full recalculation occurs, then above all rows first get deleted and then recreated. created_at will be greater than last_calculated_at for all of them.

Reply:
Aha. That's fair if a bit tricky. Good chat, go get this merged!